Clean up config checks / fix heartbeat NPE

andrewvc opened the pull request:
This fixes a bug where heartbeat would encounter an NPE while attempting to free resources during a config check and cleans up some internal interfaces.

This change lets beats choose whether their factories implement a special config check function. If not, it is assumed their creation function will return an error and that this can be used for checking the config. It is further assumed that creation with the factory is fully GC-able, which is not the case for heartbeat, hence the need for a custom `CheckConfig` method.

elasticmachine commented:
Pinging @elastic/uptime

andrewvc commented on heartbeat/monitors/monitor_test.go:
```diff
@@ -90,12 +91,15 @@ func TestDuplicateMonitorIDs(t *testing.T) {
+
 	m1, m1Err := makeTestMon()
-	assert.NoError(t, m1Err)
+	require.NoError(t, m1Err)
```
These are now require because subsequent assertions make no sense if prior ones fail.

andrewvc commented:
This also fixes a libbeat bug where the dynamic config reloading code would start the runners specified in the factory as part of the check, but never stop them. This uses the proper `Check` method which should allow cleanup.

I tried adding a test for this, but the code is quite hard to mock and stub. It probably needs a refactor beyond the scope of this PR to become more testable.

ruflin commented on heartbeat/monitors/factory.go:
```diff
@@ -44,5 +46,19 @@ func (f *RunnerFactory) Create(p beat.Pipeline, c *common.Config, meta *common.M
-	return checkMonitorConfig(config, globalPluginsReg, f.allowWatches)
+	m, err := newMonitor(config, globalPluginsReg, nil, nil, f.allowWatches, nil)
+	// We need to free up any resources like unique ID checks
+	fmt.Printf("CHECK FREE\n")
```
Leftover? Also below?

ruflin commented on heartbeat/monitors/monitor.go:
```diff
@@ -92,13 +86,31 @@ func (e ErrDuplicateMonitorID) Error() string {
 	return fmt.Sprintf("monitor ID %s is configured for multiple monitors! IDs must be unique values.", e.ID)
 }
 
+// Creates a new monitor, without leaking resources in the event of an error.
```
I wonder why hound did not complain here.

andrewvc replied:
My guess is because it isn't public. Fixed!

ruflin commented on libbeat/cfgfile/reload.go:
```diff
@@ -111,18 +112,21 @@ func NewReloader(pipeline beat.Pipeline, cfg *common.Config) *Reloader {
 // Check configs are valid (only if reload is disabled)
 func (rl *Reloader) Check(runnerFactory RunnerFactory) error {
 	// If config reload is enabled we ignore errors (as they may be fixed afterwards)
+	debug.PrintStack()
```
Leftovers? Also below?

ruflin commented on libbeat/cfgfile/reload.go:
```diff
@@ -131,13 +135,15 @@ func (rl *Reloader) Check(runnerFactory RunnerFactory) error {
 			continue
 		}
-		_, err := runnerFactory.Create(rl.pipeline, c.Config, c.Meta)
+		err := runnerFactory.CheckConfig(c.Config)
```
Does this potentially break other Beats using reload?

andrewvc replied:
Sadly, looking at the last CI run I fear the answer is yes.

However, the current behavior breaks `heartbeat`. The problem is that there's no corresponding stop called so if `Create` leaks any resources they'll never be recovered.

I've removed the old code and manually invoked `Stop()`. I do think we should actually use the `CheckConfig` interface instead of doing a `Create` then `Stop`, but for the scope of the PR this is more appropriate.

I will open a separate PR with that refactor.

andrewvc commented:
@ruflin should be ready for another pass, I think I've addressed all your concerns.

ruflin commented on heartbeat/monitors/factory.go:
```diff
@@ -44,5 +44,17 @@ func (f *RunnerFactory) Create(p beat.Pipeline, c *common.Config, meta *common.M
+		}
+	}()
+
+	if err != nil {
```
Use just `return err`, no additional check needed.

ruflin commented on libbeat/cfgfile/reload.go:
```diff
@@ -118,11 +119,13 @@ func (rl *Reloader) Check(runnerFactory RunnerFactory) error {
 	debugf("Checking module configs from: %s", rl.path)
 	gw := NewGlobWatcher(rl.path)
 
+	debug.PrintStack()
```
Should be removed I think?

ruflin commented on libbeat/cfgfile/reload.go:
```diff
@@ -118,11 +119,13 @@ func (rl *Reloader) Check(runnerFactory RunnerFactory) error {
 		return errors.Wrap(err, "fetching config files")
 	}
 
+	debug.PrintStack()
```
Should be removed I think?

ruflin commented on libbeat/cfgfile/reload.go:
```diff
@@ -137,7 +140,8 @@ func (rl *Reloader) Check(runnerFactory RunnerFactory) error {
 		}
-		_, err := runnerFactory.Create(rl.pipeline, c.Config, c.Meta)
+		instance, err := runnerFactory.Create(rl.pipeline, c.Config, c.Meta)
+		instance.Stop()
```
Do you know what the affect here on other Beats are? I'm still concerned we might break something with it.

ruflin commented on libbeat/cfgfile/reload.go:
```diff
@@ -137,7 +137,10 @@ func (rl *Reloader) Check(runnerFactory RunnerFactory) error {
-		_, err := runnerFactory.Create(rl.pipeline, c.Config, c.Meta)
+		runner, err := runnerFactory.Create(rl.pipeline, c.Config, c.Meta)
+		if runner != nil {
+			runner.Stop()
```
I went through the different Create and Stop functions of the runners. I found a potential issue here in Metricbeat: https://github.com/elastic/beats/blob/6.7/metricbeat/mb/module/runner.go#L92

When the runner is stopped, it removes the module from the module List. As Start is never called, it means the entry was never added to the list. We have a similar issue on filebeat side: https://github.com/elastic/beats/blob/master/filebeat/fileset/factory.go#L176

Perhaps the best short term fix is to follow the other suggestion @andrewvc had to start using `CheckConfig` here but also pass pipeline and metadata, so in FB and MB the method can call Create?

andrewvc replied:
@ruflin I agree. Unfortunately, after starting work on that refactor I found that autodiscover uses `CheckConfig` and has no pipeline instance to pass it here: https://github.com/elastic/beats/blob/master/libbeat/autodiscover/autodiscover.go#L193

I propose we pause here and work on the real fix in https://github.com/elastic/beats/issues/11313

The hacks are piled too high here, and we should just do the right thing even if it's painful.

ruflin replied:
++

andrewvc commented:
@ruflin implemented a totally new approach in 7501885 . This wipes out much of the old work, except what was in the heartbeat folder. We now make the implementation of `CheckConfig` optional. Only heartbeat implements it at the moment.

ruflin commented:
I like this change as it is it removes the non implemented methods and allows Heartbeat to move forward with the interface.

andrewvc commented:
@ruflin fixed merge conflicts, updated the issue description / title (so that it will make a better commit message) and added a changelog entry. Can I get a final LGTM?

ruflin commented:
jenkins, test this

ruflin commented on vendor/github.com/docker/go-connections/tlsconfig/certpool_other.go:
```diff
@@ -4,7 +4,6 @@ package tlsconfig
 
 import (
 	"crypto/x509"
```
How did these vendor changes slip in here? Also see below.

andrewvc replied:
Fixed via a new squashed commit. Something weird was going on with git, but squashing fixed it. I couldn't remove them in subsequent commits weirdly.

ruflin commented:
@andrewvc The travis build was failing for Heartbeat, but I don't think related at all. I restarted it .

andrewvc commented:
Merging. Failures are unrelated. Metricbeat tests keep failing for different reasons. Out of three tries, two for package management servers being down.
